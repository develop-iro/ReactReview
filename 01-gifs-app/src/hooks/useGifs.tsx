import { useRef, useState } from "react";
import { contains, formatValue, isEmpty } from "../utils/utils.helper";
import { getGif } from "../data";
import type { Gif } from "../data/interfaces";

// const gifsCache: Record<string, Gif[]> = {};

export function useGifs() {
  const [gifResponse, setGifResponse] = useState<Gif[]>([]);
  const [previousSearches, setPreviousSearches] = useState<string[]>([]);
  const gifCache = useRef<Record<string, Gif[]>>({});

  const handleSearch = async (value: string, previous?: boolean) => {
    const isDuplicate = contains(previousSearches, value);
    const isInvalid = isEmpty(value) || !value;
    const isMaxSearches = previousSearches.length >= 8;
    if ((isDuplicate && !previous) || isInvalid || isMaxSearches) return;

    const formattedValue = formatValue(value);

    if (!previous) {
      setPreviousSearches((prevSearches) =>
        [value, ...prevSearches].splice(0, 8)
      );
    }

    const gifs = await getGif(
      formattedValue,
      document.documentElement.lang || "en"
    );
    setGifResponse(gifs);
    gifCache.current[value] = gifs;
  };

  const handlePreviousSearch = (search: string) => {
    const cachedGifs = gifCache.current[search];

    if (cachedGifs) {
      setGifResponse(cachedGifs);
      return;
    }

    handleSearch(search, true);
  };

  return {
    gifResponse,
    previousSearches,
    handleSearch,
    handlePreviousSearch,
  };
}
