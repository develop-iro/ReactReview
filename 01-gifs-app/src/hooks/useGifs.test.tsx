import { describe, expect, it } from "vitest";
import { useGifs } from "./useGifs";
import { renderHook } from "@testing-library/react";

describe("Hook - useGifs", () => {
  it("matches snapshot", () => {
    const { result } = renderHook(() => useGifs());

    expect(result).toMatchSnapshot();
  });
});
