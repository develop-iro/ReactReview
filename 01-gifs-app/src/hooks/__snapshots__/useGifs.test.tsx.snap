// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hook - useGifs > matches snapshot 1`] = `
{
  "current": {
    "gifResponse": [],
    "handlePreviousSearch": [Function],
    "handleSearch": [Function],
    "previousSearches": [],
  },
}
`;
